railties/lib/rails/commands/command.rb

```diff
@@ -0,0 +1,88 @@
+module Rails
+  module Commands
+    class Command
+      attr_reader :argv
+
+      def initialize(argv)
+        @argv = argv
+
+        @option_parser = build_option_parser
+        @options = {}
+      end
+
+      def run(task_name)
+        command_name = Command.name_for(task_name)
+
+        parse_options_for(command_name)
+        @option_parser.parse! @argv
+
+        if command_instance = command_instance_for(command_name)
+          command_instance.public_send(command_name)
+        else
+          puts @option_parser
+        end
+      end
+
+      def tasks
+        public_instance_methods.map { |method| method.gsub('_', ':') }
+      end
+
+      def exists?(task_name)
+        command_name = Command.name_for(task_name)
+        !command_instance_for(command_name).nil?
+      end
```

Comment:
Don't like this instance method nestled in with the other class methods. Let's put it below them.